video-streaming-app-with-directus-and-sveltekit/index.md

```diff
@@ -0,0 +1,387 @@
+    {#each video.tags as tag}
+      <span class="tag">{tag}</span>
+    {/each}
+  </div>
+{:else}
+  <p>Loading...</p>
+{/if}
+
+<style>
+  .tags {
+    display: flex;
+    flex-wrap: wrap;
+    gap: 0.5rem;
+  }
+  .tag {
+    background-color: #eee;
+    padding: 0.25rem 0.5rem;
+    border-radius: 0.25rem;
+  }
+</style>
+```
+Now click on any of the videos to stream it. 
+
+![Video Player](<Screenshot 2024-07-04 at 12.27.56.png>)
+
+## Search Functionality
+In your `services/index.ts`, add a new funtion that implements search functionality to find videos by title or description.
+
+```ts
++
+export async function searchVideos(query: string): Promise<Video[]> {
+  const directus = getDirectusClient();
+  const response = await directus.request(
+    readItems("videos", {
+      search: query,
+      fields: ["*", "thumbnail.*", "video_file.*"],
+    })
+  );
+  return response as Video[];
+}
+```
+This function uses the `search` parameter from Directus to perform a search on `videos` collection.
+
+Update the code in your `routes/+page.svelte` file to use the `searchVideos` function to add search functionality to your page.
+
+```svelte
++
+<script lang="ts">
+  import { onMount } from "svelte";
+  import { getVideos, searchVideos } from "$lib/services/index";
+  import VideoGrid from "$lib/components/VideoGrid.svelte";
+  import type { Video } from "$lib/types";
+
+  let videos: Video[] = [];
+  let searchQuery = "";
+  let searchResults: Video[] = [];
+  let isSearching = false;
+
+  onMount(async () => {
+    await loadLatestVideos();
+  });
+
+  async function loadLatestVideos() {
+    try {
+      videos = (await getVideos({
+        sort: ["-upload_date"],
+        limit: 20,
+        fields: ["*", "thumbnail.*", "video_file.*"],
+      })) as Video[];
+    } catch (error) {
+      console.error("Error fetching videos:", error);
+    }
+  }
+
+  async function handleSearch() {
+    if (searchQuery.trim()) {
+      isSearching = true;
+      try {
+        const response = await searchVideos(searchQuery);
+        searchResults = response as Video[];
+      } catch (error) {
+        console.error("Error searching videos:", error);
+      } finally {
+        isSearching = false;
+      }
+    } else {
+      searchResults = [];
+    }
+  }
+</script>
+
+<h1>Stream your favourite vidoes</h1>
+
+<form on:submit|preventDefault={handleSearch}>
+  <input type="text" bind:value={searchQuery} placeholder="Search for videos" />
+  <button type="submit">Search</button>
+</form>
+
+{#if isSearching}
+  <p>Searching...</p>
+{:else if searchResults.length > 0}
+  <h2>Search Results</h2>
+  <VideoGrid videos={searchResults} />
+{:else if searchQuery}
+  <p>No results found.</p>
+{:else}
+  <h2>Latest Videos</h2>
+  {#if videos.length > 0}
+    <VideoGrid {videos} />
+  {:else}
+    <p>Loading videos...</p>
+  {/if}
+{/if}
+```
+
+You can now search and stream any video of your choice.
+![Search videos](<Screenshot 2024-07-04 at 12.48.17.png>)
```

Comment:
````suggestion
## Implementing Search

In your `services/index.ts`, add a new funtion that implements search functionality to find videos by title or description.

```ts
export async function searchVideos(query: string): Promise<Video[]> {
  const directus = getDirectusClient();
  const response = await directus.request(
    readItems("videos", {
      search: query,
      fields: ["*", "thumbnail.*", "video_file.*"],
    })
  );
  return response as Video[];
}
```

This function uses the `search` parameter from Directus to perform a search on `videos` collection.

Update the code in your `routes/+page.svelte` file to use the `searchVideos` function to add search functionality to your page.

```svelte
<script lang="ts">
  import { onMount } from "svelte";
  import { getVideos, searchVideos } from "$lib/services/index";
  import VideoGrid from "$lib/components/VideoGrid.svelte";
  import type { Video } from "$lib/types";

  let videos: Video[] = [];
  let searchQuery = "";
  let searchResults: Video[] = [];
  let isSearching = false;

  onMount(async () => {
    await loadLatestVideos();
  });

  async function loadLatestVideos() {
    try {
      videos = (await getVideos({
        sort: ["-upload_date"],
        limit: 20,
        fields: ["*", "thumbnail.*", "video_file.*"],
      })) as Video[];
    } catch (error) {
      console.error("Error fetching videos:", error);
    }
  }

  async function handleSearch() {
    if (searchQuery.trim()) {
      isSearching = true;
      try {
        const response = await searchVideos(searchQuery);
        searchResults = response as Video[];
      } catch (error) {
        console.error("Error searching videos:", error);
      } finally {
        isSearching = false;
      }
    } else {
      searchResults = [];
    }
  }
</script>

<h1>Stream your favorite videos</h1>

<form on:submit|preventDefault={handleSearch}>
  <input type="text" bind:value={searchQuery} placeholder="Search for videos" />
  <button type="submit">Search</button>
</form>

{#if isSearching}
  <p>Searching...</p>
{:else if searchResults.length > 0}
  <h2>Search Results</h2>
  <VideoGrid videos={searchResults} />
{:else if searchQuery}
  <p>No results found.</p>
{:else}
  <h2>Latest Videos</h2>
  {#if videos.length > 0}
    <VideoGrid {videos} />
  {:else}
    <p>Loading videos...</p>
  {/if}
{/if}
```

You can now search and stream any video of your choice.
![Search videos](<Screenshot 2024-07-04 at 12.48.17.png>)
````

Please use the `// [!code ++]` at the end of any line you want to add a + diff to in the rendered post.

Reply:
Done